en/Building_a_Simple_Engine/Mobile_Development/03_performance_optimizations.adoc

```diff
@@ -0,0 +1,375 @@
+
+    if (!pixels) {
+        throw std::runtime_error("Failed to load texture image: " + path);
+    }
+
+    // Check if the dimensions are power of two
+    bool is_pot = (width & (width - 1)) == 0 && (height & (height - 1)) == 0;
+
+    if (!is_pot) {
+        // Resize to power of two
+        vk::Extent2D pot_size = make_power_of_two(width, height);
+
+        stbi_uc* resized_pixels = new stbi_uc[pot_size.width * pot_size.height * 4];
+        stbir_resize_uint8(pixels, width, height, 0,
+                          resized_pixels, pot_size.width, pot_size.height, 0, 4);
+
+        stbi_image_free(pixels);
+        pixels = resized_pixels;
+        width = pot_size.width;
+        height = pot_size.height;
+    }
+
+    // Create Vulkan image and upload pixels
+    // ...
+
+    // Free the pixel data
+    stbi_image_free(pixels);
+}
+----
+
+==== Efficient Texture Formats
+
+Choosing the right texture format is crucial for mobile performance:
+
+1. *Compressed Formats*: Use hardware-supported compressed formats whenever possible:
+   - *ASTC* (Adaptive Scalable Texture Compression): Widely supported on modern mobile GPUs, offers excellent quality-to-size ratio with flexible block sizes.
+   - *ETC2/EAC*: Required for OpenGL ES 3.0 and supported by most Android devices.
+   - *PVRTC*: Primarily supported on iOS devices with PowerVR GPUs.
+   - *BC* (Block Compression): More common on desktop but supported by some high-end mobile GPUs.
+
+2. *Format Selection Based on Content*: Choose formats based on the type of texture:
+   - Use ASTC 4x4 or 6x6 for normal maps and detailed textures.
+   - Use ASTC 8x8 or 12x12 for diffuse/albedo textures.
+   - Consider using R8 for single-channel textures like height maps.
+
+Here's how to check for and use compressed formats in Vulkan:
+
+[source,cpp]
+----
+bool is_format_supported(vk::PhysicalDevice physical_device, vk::Format format, vk::ImageTiling tiling,
+                        vk::FormatFeatureFlags features) {
+    vk::FormatProperties props = physical_device.getFormatProperties(format);
+
+    if (tiling == vk::ImageTiling::eLinear) {
+        return (props.linearTilingFeatures & features) == features;
+    } else if (tiling == vk::ImageTiling::eOptimal) {
+        return (props.optimalTilingFeatures & features) == features;
+    }
+
+    return false;
+}
+
+vk::Format find_supported_format(vk::PhysicalDevice physical_device,
+                               const std::vector<vk::Format>& candidates,
+                               vk::ImageTiling tiling,
+                               vk::FormatFeatureFlags features) {
+    for (vk::Format format : candidates) {
+        if (is_format_supported(physical_device, format, tiling, features)) {
+            return format;
+        }
+    }
+
+    throw std::runtime_error("Failed to find supported format");
+}
+
+vk::Format find_best_compressed_format(vk::PhysicalDevice physical_device) {
+    // Try ASTC formats first (best quality-to-size ratio)
```

Comment:
I'm not sure this makes sense. You will typically have created your compressed assets up-front, in which case the format is given.

Reply:
That might be application dependent.  However, you're right, there's little use-case for a game engine to support multiple formats, and even less of a reason for a tutorial on the subject to demonstrate that kind of feature.  I'll reword that, thanks!

Reply:
Removed it and stated that most engines standard practice wouldn't be to support formats other than what is created up-front.